src/components/svg-wrapper.tsx

```diff
@@ -151,6 +185,26 @@ const SvgWrapper = () => {
     const handleBackgroundUp = useEvent((e: React.PointerEvent<SVGSVGElement>) => {
         // when user holding the shift key and mis-click the background
         // preserve the current selection
+        if (mode === 'select') {
+            const { x, y } = getMousePosition(e);
+            // console.info(
+            //     selectStart,
+            //     (x * svgViewBoxZoom) / 100 + svgViewBoxMin.x,
+            //     (y * svgViewBoxZoom) / 100 + svgViewBoxMin.y
+            // );
+            HandleSelectTool(
+                graph.current,
+                selectStart.x,
+                selectStart.y,
+                (x * svgViewBoxZoom) / 100 + svgViewBoxMin.x,
+                (y * svgViewBoxZoom) / 100 + svgViewBoxMin.y
+            ).forEach(node => {
+                dispatch(addSelected(node));
```

Comment:
there is a setSelected action where you could set selected in batch instead of one by one :)

Reply:
Use `.forEach` can make sure that the nodes are **add** to `selected` continuously rather than clear `selected` and put nodes into it.

Reply:
Doesn't `handleSelectTool` return all the nodes we want to select? Or does we support first clicking on some nodes with shift and then add more nodes to `selected` by dragging?

Reply:
HandleSelectTool will return the nodes that we select this time. 

I prefer to select some *group* of nodes without pressing the shift key.

Reply:
Then if the only scenario we support is to drag to select nodes in the rectangle, why can't we add them in batch via setSelected?

Reply:
I want to keep the nodes that I have selected last time and then add nodes I select this time.

Reply:
Okay so I can drag and select nodes 1,2,3... and then drag and select nodes a,b,c... with 1,2,3... still being selected. This could be a feature but I'm wondering if other tools behave like this too. At least Illustrator won't keep previously selected elements.

Reply:
I know that AutoCAD and Solidworks work like this.

Reply:
I'll clear the `selected` if the shift key is not pressed.

Reply:
I will use addSelected rather than setSelected, because I need make sure the node won't appear twice in `selected`.

Reply:
Using `addSelected` will be O(nm) where n is the size of the current selected and m is the size of nodes returned from `findNodesInRectangle`. The following code should reduce complexity to O(n+m) by creating a set.

```tsx
            const nodesInRectangle = findNodesInRectangle(
                graph.current,
                selectStart.x,
                selectStart.y,
                (x * svgViewBoxZoom) / 100 + svgViewBoxMin.x,
                (y * svgViewBoxZoom) / 100 + svgViewBoxMin.y
            );
            const allSelected = new Set([...selected, ...nodesInRectangle]);
            dispatch(setSelected([...allSelected]));
```

Reply:
Oh, I'm not good at object-oriented operations. So, is it $\Theta(n)$ to create a set from an array by `= new Set(simpleArray)`?

Reply:
And I think the time complexity of an insert operation in a set is $O(\log n)$.

Reply:
Adding one element to a set should be O(1) in JavaScript, at least in v8, as hast table is used as the internal data structure. https://stackoverflow.com/questions/33611509/es6-map-and-set-complexity-v8-implementation

Anyway, even if it is O(logn), the suggested code should be faster than filtering out duplicates using `array.prototype.includes`. That must be O(n) and is only provided as a handy function for a single add, as redux does not allow set, non-serializable objects to be put in.